### PR TITLE
add support for security groups

### PR DESCRIPTION
**Pull request by andreaturli**
if ok, I'd like to backport it to 1.9.x

### CODE REVIEW

File: providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java [hunk 1]
```diff
@@ -142,9 +146,18 @@
...
       List<String> ports = simplifyPorts(options.getInboundPorts());
-      if (ports != null){
+      if (ports != null) {
          tags.add(naming.name(ports));
       }
+      // Add tags coming from firewalls (aka securityGroups)
+      if (!options.getGroups().isEmpty()) {
+         for (String firewallName : options.getGroups()) {
+            Firewall firewall = api.firewalls().get(firewallName);
+            validateFirewall(firewall, network);
```
Comment (nacx) [inline providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java]: This is redundant, as it has already been validated in the `CreateNodesWithGroupEncodedIntoNameThenAddToSet`.

File: providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java [hunk 2]
```diff
@@ -142,9 +146,18 @@
...
+            if (!firewall.targetTags().isEmpty()) {
+               tags.addAll(firewall.targetTags());
```
Comment (nacx) [inline providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java]: Missing closing `}`.

**Comment by nacx** [comment]
Having a closer look, the strategy that validates the firewalls will call the api only once for each firewall, but then the code in the adapter will be executed once for each firewall and node being deployed, which is too many redundant calls.

Given that the adapter [already sets the configured tags](https://github.com/jclouds/jclouds/blob/master/providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/GoogleComputeEngineServiceAdapter.java#L140), I'd change the strategy to, instead of just validating, add the tags to the templateOptions.tags, so the adapter always gets the list of tags to apply properly populated. This way we eliminate all code from the adapter that is coupled to the firewall naming generation and keep everything in one place (and we save all those redundant calls). WDYT?

File: providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java [hunk 3]
```diff
@@ -186,6 +201,15 @@ private void getOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOpt
          operationDone.apply(operation);
          checkState(operation.get().httpErrorStatusCode() == null, "Could not insert firewall, operation failed %s",
                operation);
+
+         // Add tags for firewalls
+         tags.add(name);
+      }
+   }
+
+   public static void validateFirewall(Firewall firewall, URI networkUri) {
```
Comment (nacx) [inline providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java]: Does it make more sense for the signature to be `validateFirewall(Firewall firewall, Network network)`?

**Comment by nacx** [comment]
Just the minor cosmetic comment about the signature of the validate firewall method. But regardless of that, LGTM!
Good to emrge @andreaturli! Thanks!

**Comment by andreaturli** [comment]
thanks @nacx, can't see the cosmetic comment btw

- is it fine to backport this to 1.9.x ?

File: providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java [hunk 4]
```diff
@@ -186,6 +202,16 @@ private void getOrCreateFirewalls(GoogleComputeEngineTemplateOptions templateOpt
...
          checkState(operation.get().httpErrorStatusCode() == null, "Could not insert firewall, operation failed %s",
                operation);
+
+         // Add tags for firewalls
+         tags.add(name);
+      }
+      templateOptions.tags(tags);
+   }
+
+   public static void validateFirewall(Firewall firewall, URI networkUri) {
```
Comment (nacx) [inline providers/google-compute-engine/src/main/java/org/jclouds/googlecomputeengine/compute/strategy/CreateNodesWithGroupEncodedIntoNameThenAddToSet.java]: Does it make more sense for the signature to be validateFirewall(Firewall firewall, Network network)?

**Comment by nacx** [comment]
It does not introduce breaking changes, so +1 to backport it.
I've re-posted the comment.

**Comment by andreaturli** [comment]
ah! it is not easy to backport it as in 1.9.x it was on jclouds-labs, so I need to open another PR referencing this review.

**Comment by andreaturli** [comment]
rebuild please

**Comment by andreaturli** [comment]
the [error](https://jclouds.ci.cloudbees.com/job/jclouds-pull-requests/2368/console) doesn't seem related to this PR

**Comment by andreaturli** [comment]
Merged *only* at [master](https://git1-us-west.apache.org/repos/asf?p=jclouds.git;a=commit;h=deeedd55)

**Comment by nacx** [comment]
Tests fail due to this change (the order of the instance tags has been altered). Could you fix it please?
https://jclouds.ci.cloudbees.com/job/jclouds/3457/

**Comment by andreaturli** [comment]
Sorry I will asap

**Comment by andreaturli** [comment]
https://jclouds.ci.cloudbees.com/job/jclouds/org.apache.jclouds.provider$google-compute-engine/3458/changes

**Pull request closed without merging**